text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,345 @@
+
+The `package.workspace` can be omitted if it would only contain `../` (or some
+repetition of it). That is, if the root of a workspace is hierarchically the
+first `Cargo.toml` with `[workspace]` above a crate in the filesystem, then that
+crate can omit the `package.workspace` key.
+
+Next, a crate which specifies `[workspace]` **without a `members` key** will
+transitively crawl `path` dependencies to fill in this key. This way all `path`
+dependencies (and recursively their own `path` dependencies) will inherently
+become the default value for `workspace.members`.
+
+Note that these implicit relations will be subject to the same validations
+mentioned above for all of the explicit configuration as well.
+
+### Workspaces in practice
+
+Many Rust projects today already have `Cargo.toml` at the root of a repository,
+and with the small addition of `[workspace]` in the root a workspace will be
+ready for all crates in that repository. For example:
+
+* An FFI crate with a sub-crate for FFI bindings
+
+  ```
+  Cargo.toml
+  src/
+  foo-sys/
+    Cargo.toml
+    src/
+  ```
+
+* A crate with multiple in-tree dependencies
+
+  ```
+  Cargo.toml
+  src/
+  dep1/
+    Cargo.toml
+    src/
+  dep2/
+    Cargo.toml
+    src/
+  ```
+
+Some examples of layouts that will require extra configuration, along with the
+configuration necessary, are:
+
+* Trees without any root crate
+
+  ```
+  crate1/
+    Cargo.toml
+    src/
+  crate2/
+    Cargo.toml
+    src/
+  crate3/
+    Cargo.toml
+    src/
+  ```
+
+  these crates can all join the same workspace via a `Cargo.toml` file at the
+  root looking like:
+
+  ```toml
+  [workspace]
+  members = ["crate1", "crate2", "crate3"]
+  ```
+
+* Trees with multiple workspaces
+
+  ```
+  ws1/
+    crate1/
+      Cargo.toml
+      src/
+    crate2/
+      Cargo.toml
+      src/
+  ws2/
+    Cargo.toml
+    src/
+    crate3/
+      Cargo.toml
+      src/
+  ```
+
+  The two workspaces here can be configured by placing the following in the
+  manifests:
+
+  ```toml
+  # ws1/Cargo.toml
+  [workspace]
+  members = ["crate1", "crate2"]
+  ```
+
+  ```toml
+  # ws2/Cargo.toml
+  [workspace]
+  ```
+
+* Trees with non-hierarchical workspaces
+
+  ```
+  root/
+    Cargo.toml
+    src/
+  crates/
+    crate1/
+      Cargo.toml
+      src/
+    crate2/
+      Cargo.toml
+      src/
+  ```
+
+  The workspace here can be configured by placing the following in the
+  manifests:
+
+  ```toml
+  # root/Cargo.toml
+  #
+  # Note that `members` aren't necessary if these are otherwise path
+  # dependencies.
+  [workspace]
+  members = ["../crates/crate1", "../crates/crate2"]
+  ```
+
+  ```toml
+  # crates/crate1/Cargo.toml
+  [package]
+  workspace = "../root"
+  ```
+
+  ```toml
+  # crates/crate2/Cargo.toml
+  [package]
+  workspace = "../root"
+  ```
+
+Projects like the compiler will likely need exhaustively explicit configuration.
+The `rust` repo conceptually has two workspaces, the standard library and the
```

Comment:
It's somewhat unrelated to this RFC, unfortunately, bit the gist of it is that we want crates.io deps to be part of the compiler but they do not explicitly depend on the standard library, so they need to be built in two phases.